cloudbridge/cloud/providers/gce/resources.py

```diff
@@ -1063,3 +1068,253 @@ def delete(self):
+                            project=self._provider.project_name,
+                            zone=self._provider.default_zone,
+                            instance=instance_name,
+                            body=attach_disk_body).execute())
+
+    def detach(self, force=False):
+        """
+        Detach this volume from an instance.
+        """
+        if not self.attachments:
+            return
+        instance_data = self._provider.get_gce_resource_data(
+            self.attachments.instance_id)
+        if 'disks' not in instance_data:
+            return
+        device_name = None
+        for disk in instance_data['disks']:
+            if ('source' in disk and 'deviceName' in disk and
+                disk['source'] == self.id):
+                device_name = disk['deviceName']
+        if not device_name:
+            return
+        response = (self._provider.gce_compute
+                        .instances()
+                        .detachDisk(
+                            project=self._provider.project_name,
+                            zone=self._provider.default_zone,
+                            instance=instance_data.get('name'),
+                            deviceName=device_name).execute())
+
+    def create_snapshot(self, name, description=None):
+        """
+        Create a snapshot of this Volume.
+        """
+        return self._provider.block_store.snapshots.create(
+            name, self, description)
+
+    def delete(self):
+        """
+        Delete this volume.
+        """
+        response = (self._provider.gce_compute
+                        .disks()
+                        .delete(
+                            project=self._provider.project_name,
+                            zone=self._provider.default_zone,
+                            disk=self.name).execute())
+
+    @property
+    def state(self):
+        return GCEVolume.VOLUME_STATE_MAP.get(
+            self._volume.get('status'), VolumeState.UNKNOWN)
+
+    def refresh(self):
+        """
+        Refreshes the state of this volume by re-querying the cloud provider
+        for its latest state.
+        """
+        self._volume = self._provider.get_gce_resource_data(
+            self._volume.get('selfLink'))
+
+
+class GCESnapshot(BaseSnapshot):
+
+    SNAPSHOT_STATE_MAP = {
+        'PENDING': SnapshotState.PENDING,
+        'READY': SnapshotState.AVAILABLE,
+    }
+
+    def __init__(self, provider, snapshot):
+        super(GCESnapshot, self).__init__(provider)
+        self._snapshot = snapshot
+
+    @property
+    def id(self):
+        return self._snapshot.get('selfLink')
+
+    @property
+    def name(self):
+        """
+        Get the snapshot name.
+         """
+        return self._snapshot.get('name')
+
+    @name.setter
+    # pylint:disable=arguments-differ
+    def name(self, value):
+        """
+        Set the snapshot name.
+        """
+        raise NotImplementedError('Not supported by this provider.')
```

Comment:
@baizhang @mbookman Same as above, perhaps we can simulate a name and description using tags?